File: app_pojavlauncher/src/main/java/net/kdt/pojavlaunch/prefs/screens/LauncherPreferenceJavaFragment.java
package net.kdt.pojavlaunch.prefs.screens;

import static net.kdt.pojavlaunch.Architecture.is32BitsDevice;
import static net.kdt.pojavlaunch.Tools.getTotalDeviceMemory;

import android.net.Uri;
import android.os.Bundle;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.preference.EditTextPreference;

import git.artdeell.mojo.R;
import net.kdt.pojavlaunch.PojavApplication;
import net.kdt.pojavlaunch.Tools;
import net.kdt.pojavlaunch.contracts.OpenDocumentWithExtension;
import net.kdt.pojavlaunch.multirt.MultiRTConfigDialog;
import net.kdt.pojavlaunch.prefs.CustomSeekBarPreference;
import net.kdt.pojavlaunch.prefs.LauncherPreferences;
import net.kdt.pojavlaunch.utils.jre.DotnetAssets;

public class LauncherPreferenceJavaFragment extends LauncherPreferenceFragment {
    private MultiRTConfigDialog mDialogScreen;
    private final ActivityResultLauncher<Object> mVmInstallLauncher =
            registerForActivityResult(new OpenDocumentWithExtension("xz"), (data)->{
                if(data != null) Tools.installRuntimeFromUri(getContext(), data);
            });
    private final ActivityResultLauncher<String[]> mDotnetRuntimeLauncher =
            registerForActivityResult(new ActivityResultContracts.OpenDocument(), (Uri data)->{
                if(data != null) importDotnetRuntime(data);
            });

    @Override
    public void onCreatePreferences(Bundle b, String str) {
        int ramAllocation = LauncherPreferences.PREF_RAM_ALLOCATION;
        // Triggers a write for some reason
        addPreferencesFromResource(R.xml.pref_java);

        CustomSeekBarPreference memorySeekbar = requirePreference("allocation",
                CustomSeekBarPreference.class);

        int maxRAM;
        int deviceRam = getTotalDeviceMemory(memorySeekbar.getContext());

        if(is32BitsDevice() || deviceRam < 2048) maxRAM = Math.min(1024, deviceRam);
        else maxRAM = deviceRam - (deviceRam < 3064 ? 800 : 1024); //To have a minimum for the device to breathe

        memorySeekbar.setMaxKeepIncrement(maxRAM);
        memorySeekbar.setValue(ramAllocation);
        memorySeekbar.setSuffix(" MB");

        EditTextPreference editJVMArgs = findPreference("javaArgs");
        if (editJVMArgs != null) {
            editJVMArgs.setOnBindEditTextListener(TextView::setSingleLine);
        }

        requirePreference("install_jre").setOnPreferenceClickListener(preference->{
            openMultiRTDialog();
            return true;
        });
        requirePreference("install_dotnet_runtime").setOnPreferenceClickListener(preference->{
            mDotnetRuntimeLauncher.launch(new String[]{"*/*"});
            return true;
        });
    }

    private void importDotnetRuntime(Uri uri) {
        PojavApplication.sExecutorService.execute(() -> {
            try {
                DotnetAssets.installRuntime(getContext(), uri);
                Tools.runOnUiThread(() -> Toast.makeText(getContext(), R.string.preference_dotnet_runtime_installed, Toast.LENGTH_LONG).show());
            } catch (Exception e) {
                Tools.showError(getContext(), e);
            }
        });
    }

    private void openMultiRTDialog() {
        if (mDialogScreen == null) {
            mDialogScreen = new MultiRTConfigDialog();
            mDialogScreen.prepare(getContext(), mVmInstallLauncher);
        }
        mDialogScreen.show();
    }
}

File: app_pojavlauncher/src/main/java/net/kdt/pojavlaunch/prefs/screens/LauncherPreferenceJavaFragment.kt
package net.kdt.pojavlaunch.prefs.screens

import android.os.Bundle
import android.widget.TextView
import androidx.activity.result.ActivityResultLauncher
import androidx.preference.EditTextPreference
import git.artdeell.mojo.R
import net.kdt.pojavlaunch.Architecture
import net.kdt.pojavlaunch.Tools
import net.kdt.pojavlaunch.contracts.OpenDocumentWithExtension
import net.kdt.pojavlaunch.multirt.MultiRTConfigDialog
import net.kdt.pojavlaunch.prefs.CustomSeekBarPreference
import net.kdt.pojavlaunch.prefs.LauncherPreferences

class LauncherPreferenceJavaFragment : LauncherPreferenceFragment() {
    private var mDialogScreen: MultiRTConfigDialog? = null
    private val mVmInstallLauncher: ActivityResultLauncher<Any> =
        registerForActivityResult(OpenDocumentWithExtension("xz")) { data ->
            if (data != null) Tools.installRuntimeFromUri(context, data)
        }

    override fun onCreatePreferences(b: Bundle?, str: String?) {
        val ramAllocation = LauncherPreferences.PREF_RAM_ALLOCATION
        addPreferencesFromResource(R.xml.pref_java)

        val memorySeekbar = requirePreference("allocation", CustomSeekBarPreference::class.java)

        val maxRAM: Int
        val deviceRam = Tools.getTotalDeviceMemory(memorySeekbar.context)
        if (Architecture.is32BitsDevice() || deviceRam < 2048) maxRAM = Math.min(1024, deviceRam)
        else maxRAM = deviceRam - (if (deviceRam < 3064) 800 else 1024)

        memorySeekbar.setMaxKeepIncrement(maxRAM)
        memorySeekbar.value = ramAllocation
        memorySeekbar.suffix = " MB"

        val editJVMArgs = findPreference<EditTextPreference>("javaArgs")
        editJVMArgs?.setOnBindEditTextListener { it.isSingleLine = true }

        requirePreference("install_jre").setOnPreferenceClickListener {
            openMultiRTDialog()
            true
        }
    }

    private fun openMultiRTDialog() {
        if (mDialogScreen == null) {
            mDialogScreen = MultiRTConfigDialog()
            mDialogScreen!!.prepare(context, mVmInstallLauncher)
        }
        mDialogScreen!!.show()
    }
}
